marconi/src/Marconi/Index/AddressDatum.hs

```diff
@@ -0,0 +1,517 @@
+  -> [StorableEvent AddressDatumHandle]
+asEvents events =
+    fmap toEvent $ NonEmpty.groupWith (\(_, _, _, s, _) -> s) events
+ where
+     toEvent
+         :: NonEmpty ( C.AddressAny
+                     , C.Hash C.ScriptData
+                     , Maybe C.ScriptData
+                     , C.SlotNo
+                     , C.Hash C.BlockHeader
+                     )
+         -> StorableEvent AddressDatumHandle
+     toEvent es =
+         let (_, _, _, slot, blockHash) = NonEmpty.head es
+          in AddressDatumIndexEvent (toAddressDatums es) (toDatumMap es) (C.ChainPoint slot blockHash)
+
+     toAddressDatums
+         :: NonEmpty ( C.AddressAny
+                     , C.Hash C.ScriptData
+                     , Maybe C.ScriptData
+                     , C.SlotNo
+                     , C.Hash C.BlockHeader
+                     )
+         -> Map C.AddressAny (Set (C.Hash C.ScriptData))
+     toAddressDatums es =
+         Map.fromListWith (<>)
+            $ NonEmpty.toList
+            $ fmap (\(addr, dh, _, _, _) -> (addr, Set.singleton dh)) es
+
+     toDatumMap
+         :: NonEmpty ( C.AddressAny
+                     , C.Hash C.ScriptData
+                     , Maybe C.ScriptData
+                     , C.SlotNo
+                     , C.Hash C.BlockHeader
+                     )
+         -> Map (C.Hash C.ScriptData) C.ScriptData
+     toDatumMap es =
+         Map.fromList
+            $ catMaybes
+            $ NonEmpty.toList
+            $ fmap (\(_, dh, d, _, _) -> fmap (dh,) d) es
+
+instance Queryable AddressDatumHandle where
+  queryStorage
+    :: Foldable f
+    => QueryInterval C.ChainPoint
+    -> f (StorableEvent AddressDatumHandle)
+    -> AddressDatumHandle
+    -> StorableQuery AddressDatumHandle
+    -> IO (StorableResult AddressDatumHandle)
+  queryStorage qi es (AddressDatumHandle c _) AllAddressesQuery = do
+    persistedData :: [(C.AddressAny, C.Hash C.ScriptData, Maybe C.ScriptData, C.SlotNo, C.Hash C.BlockHeader)] <-
+      case qi of
+        QEverything ->
+            SQL.query c
+                [r|SELECT address, address_datums.datum_hash, datumhash_datum.datum, slot_no, block_hash
+                   FROM address_datums
+                   LEFT JOIN datumhash_datum
+                   ON datumhash_datum.datum_hash = address_datums.datum_hash
+                   ORDER BY slot_no ASC, address, datumhash_datum.datum_hash|]
+                ()
+        QInterval (C.ChainPoint b _) (C.ChainPoint e _) ->
+            SQL.query c
+                [r|SELECT address, address_datums.datum_hash, datumhash_datum.datum, slot_no, block_hash
+                   FROM address_datums
+                   LEFT JOIN datumhash_datum
+                   ON datumhash_datum.datum_hash = address_datums.datum_hash
+                   WHERE slot_no >= ? AND slot_no <= ?
+                   ORDER BY slot_no ASC, address, datumhash_datum.datum_hash|]
+                (b, e)
```

Comment:
This is not correct, and is actually way more complicated than one would expect. You should only have `slot_no <= e` here and here is a convoluted explanation as to why :)

The abstract layout of the data (regardless of location) is something like: `e1, e2, e3, e4, acc`. When you give a query interval, what you are saying is try to find a good `e` in that interval and then give me all the `e`s that are lower than that.

This makes sense if you think that what you need to do here is compute `acc` at `e1, e2, e3,...`, but in order to do that you will always need *all* the events lower than the one you select.

Now, the question arises: Why have an interval at all? Why not just have some slot number that everyone synchronises on? The question would be that it *might* be useful for the client to assume a certain risk which sounds like: "if the indexers are within 3 slots of each-other I am fine with absorbing the risk of inconsistent blockchain views".

Context: https://github.com/input-output-hk/plutus-apps/blob/66799bd9c15fe125bb334501f48cf3ee1d740ec3/rewindable-index/src/RewindableIndex/Storable.hs#L318

In order for this function to work properly we need to select all of the DB events which are `<= end`. In retrospect the naming for the intervals is also confusing because slot numbers always grow, and intervals are specified as (low, high) the `end` is actually the latest, highest, top slot number. I find that kind of confusing. Hopefully it's just me.

If you rewrite this function like suggested you get the added bonus of not needing the case that follows it. :)

Reply:
Changed to return an empty list instead for the time being. Will implement it correctly in a future PR once we get the ADR for the query interval.